torch/csrc/cuda/nccl.cpp

```diff
@@ -859,7 +989,11 @@ void gather(
   } else {
     NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
   }
-  NCCL_CHECK(ncclGroupEnd());
+  if (!comm_nonblocking) {
+    NCCL_CHECK(ncclGroupEnd());
+  } else {
+    NCCL_CHECK_NONBLOCKING(ncclGroupEnd(), _comm);
+  }
```

Comment:
nit: Maybe we can just pass in `comm_nonblocking` as a parameter to `NCCL_CHECK` and have the logic reside internally there instead of repeating the `if (!comm_nonblocking) {..} else {...}` pattern everywhere.

Reply:
Giving it a deeper thought, maybe we don't even need to rely on `comm_nonblocking`.
For example, let's say we define `NCCL_CHECK` like this:
```
void NCCL_CHECK(ncclResult_t rs, ncclComm_t comm) {
#if NCCL_HAS_IN_PROGRESS
  while (rs == ncclInProgress) {
    // maybe sched_yield() here?
    rs = ncclCommGetAsyncError(comm);
  }
#endif
  if (rs != ncclSuccess) throw;
}

Reply:
I think the issue is that we have three use-cases that I'm not sure can be unified cleanly:
current `NCCL_CHECK`, which doesn't take a communicator as an argument
a single nonblocking `NCCL_CHECK`, which requires a communicator as an argument
a nonblocking `NCCL_CHECK` for e.g., groupend, which would require a vector of communicators

I think the current approach is a bit verbose but very explicit, and I'm not sure hiding the existence/handling of nonblocking in `NCCL_CHECK` functions in favor of cleanliness hinders readability.

We could achieve this with overriding, but my question now is whether we would want to pass a communicator even when nonblocking is not used and it would be unnecessary. Is this kind of overhead (including the useless `ncclInProgress` check considered negligible?